crates/pessimistic-proof-utils/src/local_exit_tree.rs

```diff
@@ -0,0 +1,249 @@
+        }
+
+        LETMerkleProof { siblings }
+    }
+}
+
+impl<H, const TREE_DEPTH: usize> Into<LocalExitTree<H, TREE_DEPTH>>
+    for &LocalExitTreeData<H, TREE_DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Default + Serialize + DeserializeOwned,
+{
+    fn into(self) -> LocalExitTree<H, TREE_DEPTH> {
+        let leaf_count = self.layers[0].len();
+        let mut frontier = [H::Digest::default(); TREE_DEPTH];
+        let mut index = leaf_count;
+        let mut height = 0;
+        while index != 0 {
+            if index & 1 == 1 {
+                frontier[height] = self.layers[height][index ^ 1];
+            }
+            height += 1;
+            index >>= 1;
+        }
+
+        LocalExitTree::from_parts(
+            leaf_count
+                .try_into()
+                .expect("usize expected to be at least 32 bits"),
+            frontier,
+        )
+    }
+}
+
+impl<H, const TREE_DEPTH: usize> LETMerkleProof<H, TREE_DEPTH>
+where
+    H: Hasher,
+    H::Digest: Eq + Copy + Default + Serialize + DeserializeOwned,
+{
+    pub fn verify(&self, leaf: H::Digest, leaf_index: usize, root: H::Digest) -> bool {
+        let mut entry = leaf;
+        let mut index = leaf_index;
+        for &sibling in &self.siblings {
+            entry = if index & 1 == 0 {
+                H::merge(&entry, &sibling)
+            } else {
+                H::merge(&sibling, &entry)
+            };
+            index >>= 1;
+        }
+        if index != 0 {
+            return false;
+        }
+        if entry != root {
+            return false;
+        }
+
+        true
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use pessimistic_proof::local_exit_tree::hasher::Keccak256Hasher;
+    use pessimistic_proof::local_exit_tree::LocalExitTree;
+    use rand::{random, thread_rng, Rng};
+
+    use crate::local_exit_tree::LocalExitTreeData;
+
+    const TREE_DEPTH: usize = 32;
+    type H = Keccak256Hasher;
+
+    fn compare_let_data_let_frontier(num_leaves: usize) {
+        let leaves = (0..num_leaves).map(|_| random()).collect::<Vec<_>>();
+        let local_exit_tree_frontier: LocalExitTree<H, TREE_DEPTH> =
+            LocalExitTree::from_leaves(leaves.iter().cloned());
+        let local_exit_tree_data: LocalExitTreeData<H, TREE_DEPTH> =
+            LocalExitTreeData::from_leaves(leaves.into_iter());
+        assert_eq!(
+            local_exit_tree_frontier.get_root(),
+            local_exit_tree_data.get_root()
+        );
+    }
+
+    #[test]
+    fn test_data_vs_frontier_empty() {
+        compare_let_data_let_frontier(0)
+    }
+
+    #[test]
+    fn test_data_vs_frontier_root() {
+        let num_leaves = thread_rng().gen_range(1..100.min(1 << TREE_DEPTH));
+        compare_let_data_let_frontier(num_leaves)
+    }
+
+    #[test]
+    fn test_data_vs_frontier_add_leaf() {
+        let num_leaves = thread_rng().gen_range(1usize..100.min(1 << TREE_DEPTH));
+        let leaves = (0..num_leaves).map(|_| random()).collect::<Vec<_>>();
```

Comment:
Have you considered using something like `proptest`/`quickcheck` to generate these test cases? That gives some useful features like test case shrinking and ability to deterministically reproduce a failed test.

Reply:
I've considered using some of these testing libraries in the past, but never pulled the trigger. I agree that they could be useful though. Maybe I can do that in a separate PR, as I would need to read the docs first.